### ISSUE TITLE
Popularity

### ISSUE THREAD

**Issue by petermasking** [MaskingTechnology/comify#165]
We need to design a popularity system that we can use for the explore features.

**Comment by basmasking** [comment]
The first and simple implementation could be very simple and show the posts ordered by the most likes on the post. I think for now that is good enough.

**Comment by petermasking** [comment]
Should we also take the number of reactions into account to calculate the score based on all engagements?

For the creator popularity I think we can use the average comic post popularity score and the number of followers for the calculation.

**Comment by basmasking** [comment]
> Should we also take the number of reactions into account to calculate the score based on all engagements?

We could, but then I suggest to rate a reaction higher than a like. People tend to like easier then that they leave a reaction.

> For the creator popularity I think we can use the average comic post popularity score and the number of followers for the calculation.

I like the idea of incorporating the number of followers, but I don't know if this is going to work the way we want.

Let's look at the following scenario: creator A creates a lot of good posts. Let's say A has 100 posts with an average of 300 likes, and has 1000 followers. The popularity score for A would be 0.3 (300 / 1000). Now that the creator is ranked number 1 in the explore creator ranking, A gains 501 new followers before a new post has been created. The popularity score now drops to 0.1998 (300 / 1501).

Creator B has 1 post with of 2 stars, and has 10 followers. The popularity for B would be 0.2 (2 / 10). Now that A has gained a lot of followers, B is now the most popular creator.

What I like about this idea is that creators with a small follower base can become popular quickly if they post some good content. Having a lot of people simply clicking on the post doesn't make you popular.

What I don't like is that if you have gone viral with a single post and don't do anything anymore, you can stay popular for a long time. Especially if people start to unfollow you, because then the number of followers decreases, making you even more popular.

Should we also keep track of the activity of the creator then?

**Comment by basmasking** [comment]
I started with this issue and ran into a challenge. We need to resolve this before this story can be continued.

The reaction count, rating count, followers and following are all stored on the main domain objects itself (creator, post, reaction). We want to change this and introduce an `engagement` domain object. This separate model should help us in getting the numbers for popularity etc. easier.

**Comment by petermasking** [comment]
We've been philosophizing and playing around to come to the following model (that looks like a mystery box if you weren't there, but it's for our own reference).

![Popularity](https://github.com/MaskingTechnology/comify/assets/108156553/65a655f1-7e77-41f2-b32d-ec9d8c726b4a)

This helped us with making the decision to go for an event driven solution. The only problem is that Jitar doesn't support this yet... So we have to pick that up first before we can continue this issue. On hold for now.

**Comment by basmasking** [comment]
Now that we have finished #373 [event support], we can continue with this story.
